### PR TITLE
ApplicationAutoscaling: support autoscaling policies, deregister_scalable_target

### PR DESCRIPTION
**Pull request by jweite**
Also, changes so same ClientError is returned for validation exceptions as AWS does.

### CODE REVIEW

File: tests/test_applicationautoscaling/test_applicationautoscaling.py [hunk 1]
```diff
@@ -260,3 +262,182 @@ def test_register_scalable_target_resource_id_variations():
...
+    response = client.describe_scaling_policies(
+        PolicyNames=[policy_name],
+        ServiceNamespace=namespace,
+        ResourceId=resource_id,
+        ScalableDimension=scalable_dimension,
+    )
+    response["ResponseMetadata"]["HTTPStatusCode"].should.equal(200)
+    len(response["ScalingPolicies"]).should.equal(0)
+
+    pass
```
Comment (bblommers) [inline tests/test_applicationautoscaling/test_applicationautoscaling.py:415]: Unnecessary

File: tests/test_applicationautoscaling/test_applicationautoscaling.py [hunk 2]
```diff
@@ -260,3 +262,182 @@ def test_register_scalable_target_resource_id_variations():
...
+    policy["ScalableDimension"].should.equal(scalable_dimension)
+    policy["PolicyType"].should.equal(policy_type)
+    policy["TargetTrackingScalingPolicyConfiguration"].should.equal(policy_body)
+    policy["PolicyARN"].should.match(
+        r"arn:aws:autoscaling:.*1:scalingPolicy:.*:resource/{}/{}:policyName/{}".format(
+            namespace, resource_id, policy_name
+        )
+    )
+    policy.should.have.key("CreationTime").which.should.be.a("datetime.datetime")
+    pass
```
Comment (bblommers) [inline tests/test_applicationautoscaling/test_applicationautoscaling.py:361]: Unnecessary

File: moto/applicationautoscaling/models.py [hunk 3]
```diff
@@ -124,6 +126,100 @@ def _add_scalable_target(self, target):
             self.targets[target.scalable_dimension][target.resource_id] = target
         return target
 
+    def deregister_scalable_target(self, namespace, r_id, dimension):
+        """ Registers or updates a scalable target. """
+        if self._scalable_target_exists(r_id, dimension):
+            del self.targets[dimension][r_id]
+        else:
+            raise AWSValidationException(
```
Comment (bblommers) [inline moto/applicationautoscaling/models.py:134]: Can we get a  unit test for this error scenario?

File: moto/applicationautoscaling/models.py [hunk 4]
```diff
@@ -124,6 +126,100 @@ def _add_scalable_target(self, target):
...
+        self, policy_name, service_namespace, resource_id, scalable_dimension
+    ):
+        policy_key = FakeApplicationAutoscalingPolicy.formulate_key(
+            service_namespace, resource_id, scalable_dimension, policy_name
+        )
+        if policy_key in self.policies:
+            del self.policies[policy_key]
+            return {}
+        else:
+            raise AWSValidationException(
```
Comment (bblommers) [inline moto/applicationautoscaling/models.py:217]: And this one as well

File: moto/applicationautoscaling/models.py [hunk 5]
```diff
@@ -199,6 +295,50 @@ def update(self, **kwargs):
...
+        self.target_tracking_scaling_policy_configuration = None
+
+        if "policy_type" == "StepScaling":
+            self.step_scaling_policy_configuration = policy_body
+            self.target_tracking_scaling_policy_configuration = None
+        elif policy_type == "TargetTrackingScaling":
+            self.step_scaling_policy_configuration = None
+            self.target_tracking_scaling_policy_configuration = policy_body
+        else:
+            raise AWSValidationException(
```
Comment (bblommers) [inline moto/applicationautoscaling/models.py:319]: Same here

**Review by bblommers — CHANGES_REQUESTED** [review]
Hey @jweite - left a few comments

**Comment by jweite** [comment]
Lost track of this one, sorry!  Made the requested changes and rebased to current master.

**Review by bblommers — APPROVED** [review]
LGTM. Thanks @jweite!

**Comment by bblommers** [comment]
This is now part of moto >= 1.3.16.dev83

**Pull request merged**